File: internal/app/tunnelstatus.go
package app

import (
	"bufio"
	"context"
	"fmt"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/chenchaoyi/gtmux/internal/diag"
)

// status/tunnel.json says whether this Mac is reachable from outside, for either
// backend (openspec change `diagnostics`, design 7.2). The pairing window and doctor
// read it. Before it, the window decided "the tunnel is down" by matching phrases in
// cloudflared's log for either backend, so a Direct user's verdict came from a tunnel
// they were not running, and the Direct client published nothing at all.

// Tunnel states.
const (
	tunnelConnecting = "connecting"
	tunnelConnected  = "connected"
	tunnelDown       = "down"
)

// tunnelStatusStale: a reporter rewrites at least every 30s, so 90s without a write
// means it has stopped.
const tunnelStatusStale = 90 * time.Second

// tunnelReporter publishes one backend's state and logs only its transitions, so a
// tunnel that stays down does not write a line every probe.
//
// States: connecting until the first success; connected after a success; down after a
// failure once connected, or when a minute of connecting has not succeeded.
type tunnelReporter struct {
	mu        sync.Mutex
	backend   string
	server    string
	pairURL   string
	started   time.Time
	state     string
	lastErr   string
	lastErrAt time.Time
	lastOK    time.Time
}

// connectGrace is how long a tunnel may keep failing before it counts as down rather
// than still dialling.
const connectGrace = time.Minute

func newTunnelReporter(backend, server, pairURL string) *tunnelReporter {
	return &tunnelReporter{backend: backend, server: server, pairURL: pairURL,
		started: time.Now(), state: tunnelConnecting}
}

// report records one probe: ok, or failed with err. report(false, "") only publishes.
func (t *tunnelReporter) report(ok bool, err string) {
	t.mu.Lock()
	defer t.mu.Unlock()
	now := time.Now()
	lg := diag.For("tunnel")
	switch {
	case ok:
		if t.state != tunnelConnected {
			lg.Info("tunnel.connected", "the tunnel reaches this Mac end to end",
				"backend", t.backend, "server", t.server)
		}
		t.state, t.lastOK = tunnelConnected, now
	case err != "":
		t.lastErr, t.lastErrAt = err, now
		lg.Debug("tunnel.probe.failed", "a tunnel probe failed", "backend", t.backend, "error", err)
		goesDown := t.state == tunnelConnected ||
			(t.state == tunnelConnecting && now.Sub(t.started) > connectGrace)
		if goesDown {
			t.state = tunnelDown
			lg.Warn("tunnel.disconnected", "the tunnel does not reach this Mac",
				"backend", t.backend, "server", t.server, "error", err)
		}
	}
	detail := map[string]any{"backend": t.backend, "server": t.server, "url": t.pairURL}
	if t.lastErr != "" {
		detail["lastError"] = t.lastErr
		detail["lastErrorAt"] = t.lastErrAt.Format(time.RFC3339)
	}
	if !t.lastOK.IsZero() {
		detail["lastOkAt"] = t.lastOK.Format(time.RFC3339)
	}
	diag.Publish("tunnel", t.state, tunnelStatusStale, detail)
}

// probeHealth asks a pairing URL's /api/health directly (no proxy: the tunnel client
// does not use one, and a proxy's path to the server says nothing about the tunnel).
func probeHealth(ctx context.Context, pairURL string) error {
	tr := &http.Transport{Proxy: nil}
	hc := &http.Client{Timeout: 8 * time.Second, Transport: tr}
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, strings.TrimRight(pairURL, "/")+"/api/health", nil)
	if err != nil {
		return err
	}
	resp, err := hc.Do(req)
	if err != nil {
		return err
	}
	_ = resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("the pairing address answered HTTP %d", resp.StatusCode)
	}
	return nil
}

// watchSelfTunnel probes the Direct pairing URL every 30s (every 5s until the first
// success) for as long as the client runs, and reports each result. The probe resolves
// the name the client dials, so a network that hijacks it shows as down with the
// resolver's error, which is the truth for Direct.
func watchSelfTunnel(ctx context.Context, server string) *tunnelReporter {
	host := server
	if u, err := url.Parse(server); err == nil && u.Host != "" {
		host = u.Host
	}
	rep := newTunnelReporter("direct", host, selfTunnelPairURL(server))
	rep.report(false, "")
	go func() {
		for {
			err := probeHealth(ctx, rep.pairURL)
			if ctx.Err() != nil {
				return
			}
			if err == nil {
				rep.report(true, "")
			} else {
				rep.report(false, err.Error())
			}
			wait := 30 * time.Second
			rep.mu.Lock()
			if rep.lastOK.IsZero() {
				wait = 5 * time.Second
			}
			rep.mu.Unlock()
			select {
			case <-time.After(wait):
			case <-ctx.Done():
				return
			}
		}
	}()
	return rep
}

// cloudflaredMetricsAddr is where gtmux asks cloudflared to publish its metrics, so the
// Standard tunnel's state is read from its registered-connection count rather than from
// phrases in its log. The ports after it are cloudflared's own defaults, for a service
// installed before gtmux passed the address explicitly.
const cloudflaredMetricsAddr = "127.0.0.1:49317"

var cloudflaredMetricsFallbacks = []string{"127.0.0.1:20241", "127.0.0.1:20242", "127.0.0.1:20243", "127.0.0.1:20244", "127.0.0.1:20245"}

// haConnections reads cloudflared_tunnel_ha_connections from a Prometheus text page.
func haConnections(body string) (int, bool) {
	sc := bufio.NewScanner(strings.NewReader(body))
	for sc.Scan() {
		line := sc.Text()
		if strings.HasPrefix(line, "cloudflared_tunnel_ha_connections ") {
			f := strings.Fields(line)
			if n, err := strconv.ParseFloat(f[len(f)-1], 64); err == nil {
				return int(n), true
			}
		}
	}
	return 0, false
}

var standardReporter *tunnelReporter

// sampleStandardTunnel is serve's slow-tick step under the Standard backend: read
// cloudflared's metrics and report. No-op for any other backend.
func sampleStandardTunnel() {
	if tunnelBackend() != "standard" {
		return
	}
	if standardReporter == nil {
		standardReporter = newTunnelReporter("standard", "cloudflare", readTunnelURL())
	}
	hc := &http.Client{Timeout: 2 * time.Second, Transport: &http.Transport{Proxy: nil}}
	for _, addr := range append([]string{cloudflaredMetricsAddr}, cloudflaredMetricsFallbacks...) {
		resp, err := hc.Get("http://" + addr + "/metrics")
		if err != nil {
			continue
		}
		b := new(strings.Builder)
		_, _ = bufio.NewReader(resp.Body).WriteTo(b)
		_ = resp.Body.Close()
		if n, ok := haConnections(b.String()); ok {
			if n > 0 {
				standardReporter.report(true, "")
			} else {
				standardReporter.report(false, "cloudflared has no registered connection to the edge")
			}
			return
		}
	}
	standardReporter.report(false, "cloudflared's metrics are not reachable: is the tunnel service running?")
}

File: macapp/Sources/GtmuxBar/TunnelStatus.swift
import Foundation

/// TunnelStatus is `status/tunnel.json`, which the CLI's tunnel reporter writes for either
/// backend (openspec change `diagnostics`): whether this Mac is reachable from outside,
/// checked end to end the way the phone connects. The pairing window used to decide "the
/// tunnel is down" by matching phrases in cloudflared's log, for either backend, so a
/// Direct user's verdict came from a tunnel they were not running.
struct TunnelStatus: Equatable {
    let state: String      // "connecting" | "connected" | "down"
    let backend: String    // "direct" | "standard"
    let lastError: String
    let fresh: Bool        // written within its staleAfter; a stale status is not a fact

    static func read(now: Date = Date()) -> TunnelStatus? {
        guard let data = FileManager.default.contents(atPath: Paths.data("status/tunnel.json")) else { return nil }
        return parse(data, now: now)
    }

    /// parse decodes the file. Pure, for the tests.
    static func parse(_ data: Data, now: Date) -> TunnelStatus? {
        guard let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let state = obj["state"] as? String else { return nil }
        let detail = obj["detail"] as? [String: Any] ?? [:]
        let staleAfter = (obj["staleAfter"] as? Double) ?? Double(obj["staleAfter"] as? Int ?? 0)
        var fresh = false
        if let u = obj["updated"] as? String, let updated = parseTime(u), staleAfter > 0 {
            fresh = now.timeIntervalSince(updated) <= staleAfter
        }
        return TunnelStatus(state: state, backend: detail["backend"] as? String ?? "",
                            lastError: detail["lastError"] as? String ?? "", fresh: fresh)
    }

    /// Go writes RFC 3339 with up to nine fractional digits, or none.
    static func parseTime(_ s: String) -> Date? {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let d = f.date(from: s) { return d }
        f.formatOptions = [.withInternetDateTime]
        return f.date(from: s)
    }
}

/// ReachVerdict is what the pairing window says about its address.
enum ReachVerdict: Equatable {
    case checking
    case reachable
    /// The Mac cannot reach its own public address, but the tunnel reports itself up:
    /// the local network is in the way (a DNS that maps the name to a private address),
    /// and a phone on another network connects.
    case tunnelUpMacCannotSee
    case tunnelDown(String)
    case cannotReachYet

    /// verdict combines the window's own probe with the tunnel's status. Pure, for the
    /// tests. A status that is missing or stale says nothing, so it is never a verdict.
    static func of(probeOK: Bool?, status: TunnelStatus?) -> ReachVerdict {
        guard let ok = probeOK else { return .checking }
        if ok { return .reachable }
        guard let s = status, s.fresh else { return .cannotReachYet }
        switch s.state {
        case "connected": return .tunnelUpMacCannotSee
        case "down": return .tunnelDown(s.lastError)
        default: return .cannotReachYet
        }
    }

    /// Whether the address is NOT answering right now. A pairing window just after a move
    /// uses this to say "reconnecting" instead of "can't reach it yet", which is the same
    /// fact with the wrong conclusion drawn from it.
    var isNotReachable: Bool {
        switch self {
        case .reachable, .checking, .tunnelUpMacCannotSee: return false
        case .tunnelDown, .cannotReachYet: return true
        }
    }
}
